Add GDASapp (first wave of JEDI changes)

RussTreadon-NOAA opened the pull request:
**Description**
This PR is opened to merge changes associated with g-w issue #521 from g-w branch feature/ufsda_gdasapp into develop.   feature/ufsda_gdasapp contains g-w extensions for JEDI based atmospheric DA.   

Specifically, this PR adds the option to add JEDI based variational and/or ensemble DA jobs to replace GSI based DA jobs.   The toggling on/off of JEDI_VAR and JEDI_ENS jobs is controlled via two new variables added to `config.base.emc.dyn` and `config.base.nco.static`
```
# DA engine
export DO_JEDIVAR="NO"
export DO_JEDIENS="NO"
```
When both variables are `NO`, the global workflow uses GSI based DA jobs.   Thus, the PR does not alter the default behavior of the develop global workflow.

When `DO_JEDIVAR=YES`, GSI jobs `anal` and `analdiag` are replaced by JEDI_VAR jobs `atmanalprep`, `atmanalrun`, and `atmanalpost`.  When `DO_JEDIENS=YES`, GSI jobs `eobs`, `ediag`, and `eupd` are replaced by JEDI_ENS jobs `atmensanalprep`, `atmensanalrun`, and `atmensanalpost`.   

`checkout.sh`, `build_all.sh`, and `link_workflow.sh` are updated to clone, build, and install the GDASapp in the global workflow. 
  Local directory `sorc/gdas.cd` contains the GDASApp superstructure plus the relevant components of JEDI needed to run GDASApp.

Closes #521

**Type of change**
- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
Cycled parallels have been run on Orion using a C96C48L127 testbed over the period 21-31 December 2021.   GSI and JEDI based combinations have been tested and found to successfully run to completion.    Cycled tests have been limited to Orion because the JEDI database needed to run the test period only exists on Orion.   A clone and build test was been completed on Hera.  It is not possible to build JEDI apps on WCOSS2 since required libraries have not been installed.
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] New and existing tests pass with my changes

RussTreadon-NOAA committed:
Issue #529:  add ufsda to checkout, build, and link

RussTreadon-NOAA committed:
Issue #529:  merge branch develop at 033b1d8 into feature/ufsda

CoryMartin-NOAA committed:
atmanal rocoto job scripts

CoryMartin-NOAA committed:
Start of env and parm/config changes for atmanal

CoryMartin-NOAA committed:
Changes to parm/config files

CoryMartin-NOAA committed:
Make rocoto jobs all use same j-job for atmanal tasks

CoryMartin-NOAA committed:
start of atmanal config and added to resources

CoryMartin-NOAA committed:
Rename static B template YAML

CoryMartin-NOAA committed:
Add APRUN_ATMANAL to ORION env

CoryMartin-NOAA committed:
Additional config changes for atmanal

CoryMartin-NOAA committed:
Merge branch 'feature/ufsda-rocoto-jobs' into feature/ufsda-parm

CoryMartin-NOAA committed:
typo/bugfixes

CoryMartin-NOAA committed:
Merge branch 'feature/ufsda-rocoto-jobs' into feature/ufsda-parm

CoryMartin-NOAA committed:
reduce wallclock for atmanalpost to 1 hour

CoryMartin-NOAA committed:
change r2d2 vars

CoryMartin-NOAA committed:
make r2d2 exp oper_gdas

CoryMartin-NOAA committed:
Fix for orion env file and add link to j-job

CoryMartin-NOAA committed:
Change rocoto job variables

CoryMartin-NOAA committed:
Typo in config.atmanal

CoryMartin-NOAA committed:
config.resources bug

RussTreadon-NOAA committed:
Add link to ufsda directory

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/ufsda

RussTreadon-NOAA committed:
Issue #521:  merge branch 'develop' at 97ebc4d into feature/ufsda-rocoto-jobs

RussTreadon-NOAA committed:
Issue #532:  merge branch 'develop' at 97ebc4d into feature/ufsda-parm

RussTreadon-NOAA committed:
Merge branch 'feature/ufsda-rocoto-jobs' into feature/ufsda-parm

RussTreadon-NOAA committed:
Issue #529:  merge branch 'feature/ufsda-parm' at 74f6f11 into feature/ufsda

RussTreadon-NOAA committed:
replace atmanal with atmanalrun in ORION.env

RussTreadon-NOAA committed:
Merge branch 'develop' at f8867d3 into feature/ufsda

RussTreadon-NOAA committed:
Issue #521:  extend rocoto workflow xml generator to include UFS DA

RussTreadon-NOAA committed:
Issue #521 add missing task_prep to setup_workflow.py

RussTreadon-NOAA committed:
Merge branch develop at cb8b5ad into feature/ufsda

RussTreadon-NOAA committed:
Merge branch develop at e3d64ab into feature/ufsda

RussTreadon-NOAA committed:
Merge branch 'develop' at d758e8b into feature/ufsda

CoryMartin-NOAA committed:
Change UFS-DA to GDASApp

CoryMartin-NOAA committed:
Updated build for GDAS

CoryMartin-NOAA committed:
Changes to rocoto scripts

CoryMartin-NOAA committed:
some changes to config.atmanal

CoryMartin-NOAA committed:
Changes to link_workflow and temp change to feature branch for gdas

CoryMartin-NOAA committed:
Changes to config files

RussTreadon-NOAA committed:
Updates to cycle GDASApp

 * .gitignore - add GDASApp directories and file
 * env/ORION.env - add sfcanl, add task count to APRUN_ATMANAL
 * jobs/JGLOBAL_ATMOS_SFCANL - new j-job for surface analysis
 * jobs/rocoto/postanl.sh - new rocoto job for analysis only post
 * jobs/rocoto/sfcanl.sh - new rocoto job for surface analysis
 * parm/config/config.base.emc.dyn - replace DO_UFSDA with DO_JEDIVAR
 * parm/config/config.base.nco.static - replace DO_UFSDA with DO_JEDIVAR
 * scripts/exglobal_atmos_sfcanl.sh - new ex script for surface analysis
 * sorc/checkout.sh - check out head of NOAA-EMC/GSI develop
 * sorc/link_workflow.sh - remove version numbers from MinMon paths
 * ush/rocoto/setup_workflow.py - updates for JEDIVAR

RussTreadon-NOAA committed:
Update gsi build and link

RussTreadon-NOAA committed:
Extend prototype JEDI-GDAS cycling to include JEDI hybvar and lgetkf

RussTreadon-NOAA committed:
Merge branch develop at b41a36a into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Add jobs and configs for JEDI_VAR and JEDI_ENS

RussTreadon-NOAA committed:
Remove jobs/JGDAS_GLOBAL_ATMOS_ENSANAL (erroneously committed)

RussTreadon-NOAA committed:
Add external jobs/JGDAS_GLOBAL_ATMOS_ENSANAL to .gitignore

RussTreadon-NOAA commented:
The following PRs in other NOAA-EMC repositories are associated with this g-w PR

- [NOAA EMC/GSI PR #415](https://github.com/NOAA-EMC/GSI/pull/415)
  - This GSI PR removes global_cycle from DA jobs and scripts.   
  - This g-w PR includes the addition of global_cycle jobs and scripts.
- [NOAA EMC/GDASApp PR #66](https://github.com/NOAA-EMC/GDASApp/pull/66)
  - This GDASApp PR adds JEDI cycling components to GDASApp. 
  - This g-w PR adds g-w hooks to exercise JEDI atmospheric DA components in cycling mode.

aerorahul commented on jobs/rocoto/atmanalprep.sh:
```diff
@@ -0,0 +1,16 @@
+
+###############################################################
+# set JJOB variables
+export configs="base atmanal atmanalprep"
+export EXSCRIPT=${GDASPREPPY:-$HOMEgfs/sorc/gdas.cd/scripts/exgdas_global_atmos_analysis_prep.py}
```
These variables should not be in the rocoto job script. 
Thinking ahead for the `ecflow` equivalent, would they be in the `.ecf`?
`base` adn `atmanal` are common to the 3 `atmanalXXXX.sh` jobs.  The last one `prep` or `post` or `run` need to be passed through.
I can think on this a bit more, but wanted to put this down as a comment.

RussTreadon-NOAA replied:
You are correct.  This approach does not work with `ecflow`.    

Currently the GDASApp has a single j-job, `JGDAS_GLOBAL_ATMOS_ANALYSIS`, for atmospheric JEDI-VAR DA.   This single j-job is used for prep, run, and post with toggles placed in rocoto jobs.   

One way to refactor this is to have separate GDASApp j-jobs.  That is, we replace `JGDAS_GLOBAL_ATMOS_ANALYSIS` with

- `JGDAS_GLOBAL_ATMOS_ANALYSIS_PREP`
- `JGDAS_GLOBAL_ATMOS_ANALYSIS_RUN`
- `JGDAS_GLOBAL_ATMOS_ANALYSIS_POST`

The toggles currently in `atmanalprep.sh`, `atmanalrun.sh`, and `atmanalpost.sh` would then move into their respective j-jobs.
   
This conversation also applies to JEDI-ENS DA.   The GDASApp currently has a single j-job for JEDI-ENS DA prep, run, and post.   It, too, could be refactored as suggested above.

Thoughts, comments? @aerorahul , @CoryMartin-NOAA , @WalterKolczynski-NOAA

RussTreadon-NOAA replied:
[6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1) split the multifunction `JGDAS_GLOBAL_ATMOS_ANALYSIS` and `JGDAS_GLOBAL_ATMOS_ENSANAL` j-jobs into function specific j-jobs as outlined above.

aerorahul commented on parm/config/config.atmanal:
```diff
@@ -0,0 +1,24 @@
+#!/bin/ksh -x
```
might as well use `bash` here and all `config` files.

WalterKolczynski-NOAA replied:
Not just config, but all new scripts. I'm slowly trying to convert all shell scripts to bash, so it'd be better if more non-bash weren't added.

RussTreadon-NOAA replied:
New config files have been converted to `bash`

aerorahul commented on parm/config/config.postanl:
```diff
@@ -0,0 +1,46 @@
+export GRIBVERSION='grib2'
+export SENDCOM="YES"
+
+echo "END: config.post"
```
```suggestion
echo "END: config.postanl"
```

RussTreadon-NOAA replied:
`config.post` changed to `config.postanl`

aerorahul commented on scripts/exglobal_atmos_sfcanl.sh:
```diff
@@ -0,0 +1,300 @@
+
+##############################################################
+# Update surface fields in the FV3 restart's using global_cycle
+if [ $DOGCYCLE = "YES" ]; then
```
Is there a need for this `if` block?
If `DOGCYCLE` = `NO`, we should probably not call this job at all, right?
May be I am missing something.

RussTreadon-NOAA replied:
You are correct.   The `DOGCYCLE` `if` block has been removed and indentation adjusted.

aerorahul commented on sorc/build_gdas.sh:
```diff
@@ -0,0 +1,33 @@
+source ./machine-setup.sh > /dev/null 2>&1
+cwd=`pwd`
+
+USE_PREINST_LIBS=${USE_PREINST_LIBS:-"true"}
+if [ $USE_PREINST_LIBS = true ]; then
+  export MOD_PATH=/scratch3/NCEPDEV/nwprod/lib/modulefiles
+else
+  export MOD_PATH=${cwd}/lib/modulefiles
+fi
```
I think we removed this `MOD_PATH` business.  We can check w/ @WalterKolczynski-NOAA when he is back.

WalterKolczynski-NOAA replied:
Fairly sure we did.

RussTreadon-NOAA replied:
`MOD_PATH` removed

aerorahul commented on ush/rocoto/setup_workflow.py:
```diff
@@ -42,7 +42,7 @@ def main():
         sys.exit(1)
 
-    gfs_steps = ['prep', 'anal', 'analdiag', 'analcalc', 'gldas', 'fcst', 'postsnd', 'post', 'vrfy', 'arch']
+    gfs_steps = ['prep', 'analcalc', 'gldas', 'fcst', 'postsnd', 'postanl', 'post', 'vrfy', 'arch']
```
I will have to understand the changes in this script with the changes in the PR that consolidates the XML generation.  I can give feedback on this part separately.

RussTreadon-NOAA replied:
Agreed.  Your refactoring of the XML generation take priority.  I can work with you add JEDI pieces into the refactored XML setup scripts.

aerorahul commented on jobs/rocoto/atmanalprep.sh:
```diff
@@ -0,0 +1,16 @@
+###############################################################
+# set JJOB variables
+export configs="base atmanal atmanalprep"
+export EXSCRIPT=${GDASPREPPY:-$HOMEgfs/sorc/gdas.cd/scripts/exgdas_global_atmos_analysis_prep.py}
```
The `EXSCRIPT` variable should be defined in `config.atmanlprep`.
I am still thinking more on the `configs` and what to do here.

RussTreadon-NOAA replied:
Agreed.  See previous comment for possible resolution of this issue.

RussTreadon-NOAA replied:
[6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1) removed the `configs` and `EXSCRIPT` lines from `atmanalprep.sh`.  These lines are now in the function specific j-job, `JGDAS_GLOBAL_ATMOS_ANALYSIS_PREP`.

aerorahul commented on jobs/rocoto/atmanalrun.sh:
```diff
@@ -0,0 +1,16 @@
+###############################################################
+# set JJOB variables
+export configs="base atmanal atmanalrun"
+export EXSCRIPT=${GDASRUNSH:-$HOMEgfs/sorc/gdas.cd/scripts/exgdas_global_atmos_analysis_run.sh}
```
Should move `EXSCRIPT` to `config.atmanalrun`.

This script (`atmanalrun.sh`) is just a wrapper for rocoto similar to the `.ecf` files for ecflow.

RussTreadon-NOAA replied:
Agreed.  See suggested resolution above.

RussTreadon-NOAA replied:
Addressed via [6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1)

aerorahul commented on jobs/rocoto/atmensanalpost.sh:
```diff
@@ -0,0 +1,16 @@
+###############################################################
+# set JJOB variables
+export configs="base atmensanal atmensanalpost"
+export EXSCRIPT=${GDASPOSTPY:-$HOMEgfs/sorc/gdas.cd/scripts/exgdas_global_atmos_ensanal_post.py}
```
same comment here for `EXSCRIPT`.

RussTreadon-NOAA replied:
Agreed.  See suggested resolution above.

RussTreadon-NOAA replied:
Addressed via [6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1)

aerorahul commented on jobs/rocoto/atmensanalprep.sh:
```diff
@@ -0,0 +1,16 @@
+###############################################################
+# set JJOB variables
+export configs="base atmensanal atmensanalprep"
+export EXSCRIPT=${GDASPREPPY:-$HOMEgfs/sorc/gdas.cd/scripts/exgdas_global_atmos_analysis_prep.py}
```
here as well.

RussTreadon-NOAA replied:
Agreed.  See suggested resolution above.

RussTreadon-NOAA replied:
Addressed via [6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1)

aerorahul commented on jobs/rocoto/atmensanalrun.sh:
```diff
@@ -0,0 +1,16 @@
+###############################################################
+# set JJOB variables
+export configs="base atmensanal atmensanalrun"
+export EXSCRIPT=${GDASRUNSH:-$HOMEgfs/sorc/gdas.cd/scripts/exgdas_global_atmos_ensanal_run.sh}
```
and here.

RussTreadon-NOAA replied:
Agreed.  See suggested resolution above.

RussTreadon-NOAA replied:
Addressed via [6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1)

aerorahul commented on jobs/rocoto/postanl.sh:
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash -x
```
Curious on how this differs from `post.sh` other than the `exit 99` for `FHRGRP` that is for the forecast hours?

aerorahul replied:
https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/rocoto/post.sh

RussTreadon-NOAA replied:
You are correct.  The only difference is the lines you identify.   The UPP team should take the lead on this change.   The UPP application for the analysis differs from the forecast.   

We run executable `gfs_ncep_post -> ../sorc/upp.fd/exec/upp.x` when processing the `atmanl` and `sfcanl` files.   `upp.x` output is then used to generate analysis pgrb files.   The model runs the inline UPP.   `post.sh` does not execute `upp.x`  Instead, it processes inline UPP output to create pgrb files.  When `post.sh` includes the analysis files, the post job which processes analysis files and possibly initial forecast hours has different dependencies than subsequent post jobs which only process inline UPP output from the forecast model.   'postanl.sh` was added in recognition of these distinctions.   

Initial cycling tests with JEDI-VAR benefited from moving the analysis UPP into its own job.   Subsequent local changes to JEDI-VAR _may_ lessen the need for this.   This would need to be tested to confirm.   

That said, the above distinctions between analysis and forecast UPP applications remain.  If the UPP team were to adopt this approach, you are right in that `postanl.sh` could and should be simplified.

WalterKolczynski-NOAA replied:
If the scripts are identical, we should just use the same script. I definitely don't think it is worth it to maintain all this duplicate code for a single variable bounds check. If we really think it is necessary (I don't), we can put it in the config file.

RussTreadon-NOAA replied:
'config.postanl' and `postanl.sh` have been removed from a working copy of `feature/ufsda-gdasapp` on Hera.   References to `postanl` have been removed from `setup_workflow.py` and `config.resources`.  A four cycle GSI-based DA parallel is now running on Hera to confirm no change in the results.    A four cycle JEDI-based DA parallel will be run on Orion once the machine is returned to service.   Assuming both tests pass, `postanl` will be removed from the branch.

RussTreadon-NOAA replied:
[6674720](https://github.com/NOAA-EMC/global-workflow/commit/667472055b8db74e05467f0c7b02eb6a67bce9d1) removes all references to postanl.   This includes `jobs/rocoto/postanl.sh`, `parm/config/config.postanl`, `postanl` sections in `config.resources`, and `postanl` sections in `ush/rocoto/setup_workflow.py`

guillaumevernieres commented on .gitignore:
```diff
@@ -76,6 +76,7 @@ sorc/gfs_post.fd
 sorc/gfs_wafs.fd
 sorc/gldas.fd
 sorc/gsi.fd
+sorc/gdas.cd
```
typo?

CoryMartin-NOAA replied:
Per NCO implementation standards `.fd` indicates a directory with mainly fortran source, `.cd` with mainly C source. I made the executive decision that JEDI is 'more C than F'. :-)

RussTreadon-NOAA committed:
Update sfcanl in env and setup_workflow.py

WalterKolczynski-NOAA commented on sorc/checkout.sh:
```diff
@@ -131,7 +131,8 @@ mkdir -p ${logdir}
 errs=0
 checkout "ufs_model.fd"    "https://github.com/ufs-community/ufs-weather-model" "${ufs_model_hash:-Prototype-P8c}"; errs=$((errs + $?))
-checkout "gsi.fd"          "https://github.com/NOAA-EMC/GSI.git"                "a62dec6"                         ; errs=$((errs + $?))
+checkout "gsi.fd"          "https://github.com/NOAA-EMC/GSI.git"                "develop"                         ; errs=$((errs + $?))
```
As noted in the comment above, checkout versions should always be a *specific* version (hash or tag), not a branch.

RussTreadon-NOAA replied:
The current head of NOAA-EMC/GSI develop is [c10644e](https://github.com/NOAA-EMC/GSI/commit/c10644ee443cb7a99c4300b9037d705a49501023).   My working copy of `sorc/checkout.sh` now specifies this hash instead of develop.   I'll update this hash again if NOAA-EMC/GSI develop is updated before this PR is merged into g-w develop.

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_SFCANL:
```diff
@@ -0,0 +1,179 @@
+fi
+
+
+# Get LEVS
```
Should this be done at the script level instead? I know NCO doesn't like calculations in the j-job.

RussTreadon-NOAA replied:
Good point.   The `Get LEVS` section has been removed from `JGLOBAL_ATMOS_SFCANL`.   Script `exglobal_atmos_sfcanl.sh` already contains lines to determine `LEVS`

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_SFCANL:
```diff
@@ -0,0 +1,179 @@
+[[ $status -ne 0 ]] && exit $status
+
+
+##if [ $DOHYBVAR = "YES" ]; then
```
Please remove inactive code

RussTreadon-NOAA replied:
Inactive lines removed.

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_SFCANL:
```diff
@@ -0,0 +1,179 @@
+##fi
+
+
+# Link observational data
```
As above, would this be better at the script level?

RussTreadon-NOAA replied:
The comment is incorrect.  No files are being linked.  Furthermore, global_cycle does not use `PREPQC`, `PREPQCPF1, or `TCVITL`.   The `DONST` line has been moved to `exglobal_atmos_sfcanl.sh`.  This section of scripting has been removed from `JGLOBAL_ATMOS_SFCANL`.

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_SFCANL:
```diff
@@ -0,0 +1,179 @@
+###############################################################
+# Run relevant script
+env
+msg="HAS BEGUN on `hostname`"
```
We've removed all of the backticks from g-w. Please use `$(hostname)`.

RussTreadon-NOAA replied:
Use `$(hostname)` to replace backticks

WalterKolczynski-NOAA commented on parm/config/config.atmensanal:
```diff
@@ -0,0 +1,27 @@
+export ATMENSYAML=$HOMEgfs/sorc/gdas.cd/parm/atm/lgetkf/lgetkf.yaml
+##export FV3JEDI_FIX=/work2/noaa/da/cmartin/GDASApp/fix
+##export FV3JEDI_FIX=/work2/noaa/da/fhan/GDASApp/fix
+export FV3JEDI_FIX=/work2/noaa/da/rtreadon/GDASApp/fix
```
To match the way other fix directories are handled, this needs to be relative to the g-w clone and `link_workflow.sh` should be updated to link the directory in. Ideally, the fix directory is added into the master g-w fix location so it can be properly sync'd on all HPC, but we might be able to use the personal locations on an interim basis. Thoughts @KateFriedman-NOAA ?

KateFriedman-NOAA replied:
New fix files should go into the main fix sets. @RussTreadon-NOAA please provide the best pickup location for the new JEDI fix files and let me know the preferred subfolder name for them (`fix_jedi`?). I can pull them into the main set and then rsync to the other platform copies. Go-live on WCOSS2 is tomorrow so this may take me a day or so to get to. Thanks!

CoryMartin-NOAA replied:
Since this is still very much a 'fluid' system (and we don't yet have a good solution for background error fix files), I suggest we stick with the interim basis personal location of FV3JEDI_FIX for now, but will defer to you and @aerorahul on this matter, it will just lead to more work for EIB folks over the next year or two :-)

WalterKolczynski-NOAA replied:
Even if we go that route, I'd like all that to be handled in `link_workflow` and FV3JEDI_FIX to just be a path in the g-w tree like the others, so we keep the machine-specific paths contained.

CoryMartin-NOAA replied:
yeah that makes sense

RussTreadon-NOAA replied:
Since this will be an iterative process, I can assume (for the time being) the responsibility for maintaining a `fix_jedi` directory in the main g-w fix on Orion, Hera, and WCOSS2.   Yes, we cannot yet run JEDI on WCOSS2 but I hope we can do so before we press to far into fall 2022.   I'll add `fix_jedi` to `sorc/link_workflow.sh`.   JEDI config files will then be updated to point at `fix_jedi` in the g-w clone.  Does this work for everyone?

WalterKolczynski-NOAA replied:
@RussTreadon-NOAA Yes, Kate and I agree this is the best path for now. When fix_jedi becomes more stable, we can add it to the master fix location.

RussTreadon-NOAA replied:
@WalterKolczynski-NOAA , my preference is to add `fix_jedi` to the Hera and Orion `FIX_DIR` specified in `link_workflow.sh.   
```
elif [ $machine = "hera" ]; then
    FIX_DIR="/scratch1/NCEPDEV/global/glopara/fix_NEW"
elif [ $machine = "orion" ]; then
    FIX_DIR="/work/noaa/global/glopara/fix_NEW"
```

WalterKolczynski-NOAA replied:
@RussTreadon-NOAA Okay, coordinate that with @KateFriedman-NOAA then

WalterKolczynski-NOAA commented on parm/config/config.atmensanal:
```diff
@@ -0,0 +1,27 @@
+export R2D2_OBS_DUMP='oper_gdas'
+export R2D2_OBS_SRC='ncdiag'
+export R2D2_BC_SRC='gsi'
+##export R2D2_BC_DUMP='oper_gdas'
```
Please remove inactive code.

RussTreadon-NOAA replied:
Inactive code removed.

WalterKolczynski-NOAA commented on parm/config/config.atmensanalrun:
```diff
@@ -0,0 +1,16 @@
+
+# Task specific variables
+export JEDIENSEXE=$HOMEgfs/exec/fv3jedi_letkf.x
+##export JEDIENSEXE="/work/noaa/da/Russ.Treadon/git/global_workflow/ufsda_gdasapp/sorc/gdas.cd_pr1730/build/bin/fv3jedi_letkf.x"
```
Please remove inactive code.

RussTreadon-NOAA replied:
Inactive code removed.

WalterKolczynski-NOAA commented on parm/config/config.postanl:
```diff
@@ -0,0 +1,46 @@
+export GOESF=NO                              # goes image
+export FLXF=YES                               # grib2 flux file written by post
+
+export npe_postgp=$npe_post
```
All these resource settings should be set in `config.resources`.

RussTreadon-NOAA replied:
Resource settings moved to `config.resources`.

WalterKolczynski-NOAA commented on parm/config/config.resources:
```diff
@@ -136,6 +138,31 @@ elif [ $step = "waveawipsgridded" ]; then
+    export npe_node_atmanalprep=$(echo "$npe_node_max / $nth_atmanalprep" | bc)
+    export memory_atmanalprep="3072M"
+
+    export wtime_atmanalrun="00:30:00"
```
Is there a need for all three jobs to get the resources of the other two, or could these be separated into individual if-blocks?

RussTreadon-NOAA replied:
This section has been refactored into three if-blocks.

WalterKolczynski-NOAA commented on parm/config/config.resources:
```diff
@@ -418,6 +463,30 @@ elif [ $step = "coupled_ic" ]; then
+    export nth_atmensanalprep=1
+    export npe_node_atmensanalprep=$(echo "$npe_node_max / $nth_atmensanalprep" | bc)
+
+    export wtime_atmensanalrun="00:30:00"
```
As above, do these need to be together or can it be separated into three individual blocks?

RussTreadon-NOAA replied:
This section has be refactored into three if-blocks.

WalterKolczynski-NOAA commented on scripts/exglobal_atmos_sfcanl.sh:
```diff
@@ -0,0 +1,300 @@
+# Guess files
+GPREFIX=${GPREFIX:-""}
+GSUFFIX=${GSUFFIX:-$SUFFIX}
+ATMG03=${ATMG03:-${COMIN_GES}/${GPREFIX}atmf003${GSUFFIX}}
```
Can we condense this with a loop instead of hand-coding multiple lines doing the same thing?

RussTreadon-NOAA replied:
Examination of job log file shows that `ATM*` and `SFC*` variables are not used.   These variables plus other unused variables have been removed.

WalterKolczynski-NOAA replied:
Even better!

WalterKolczynski-NOAA commented on scripts/exglobal_atmos_sfcanl.sh:
```diff
@@ -0,0 +1,300 @@
+
+##############################################################
+# Required model guess files
+$NLN $SFCG03 sfcf03
```
As above, can we use loops to create these links instead of a bunch of hand-coded lines? I'd also love for these two sections to be combined instead of separated by 80 other lines of code so it is easier to follow what is going on.

RussTreadon-NOAA replied:
As noted above, `SFC*` variables not used and, therefore, removed.

WalterKolczynski-NOAA commented on sorc/build_all.sh:
```diff
@@ -132,6 +132,20 @@ $Build_gsi && {
+#------------------------------------
+# build gdas
+#------------------------------------
+$Build_gdas && {
```
Please use proper indentation for this block

RussTreadon-NOAA replied:
Indent to be consistent with other sections.

WalterKolczynski-NOAA commented on sorc/build_gdas.sh:
```diff
@@ -0,0 +1,33 @@
+set -eux
+
+source ./machine-setup.sh > /dev/null 2>&1
+cwd=`pwd`
```
Please replace backticks with `$(pwd)`

RussTreadon-NOAA replied:
`USE_PREINST_LIBS` section removed.  backticks around `pwd` replaced with `$(pwd)`

WalterKolczynski-NOAA commented on ush/rocoto/setup_workflow.py:
```diff
@@ -241,20 +249,30 @@ def get_gdasgfs_resources(dict_configs, cdump='gdas'):
 
-    #tasks = ['prep', 'anal', 'fcst', 'post', 'vrfy', 'arch']
-    tasks = ['prep', 'anal', 'analcalc']
+    if do_jedivar in ['Y', 'YES']:
```
Worried about how these extensive changes are going to conflict with @aerorahul's refactor of the setup_workflow scripts.

RussTreadon-NOAA replied:
Yes, these changes likely will conflict with @aerorahul 's refactor of the setup_workflow scripts.  His refactoring of the setup_workflow scripts has priority.   I'll work with him to add JEDI-DA pieces into the refactored scripts.

RussTreadon-NOAA commented:
I'll start working through comments tomorrow.   Thanks for the feedback.  This is the start of the long g-w JEDI transition.  g-w and DA will need to maintain two systems, GSI-DA and JEDI-DA, for an extended period until JEDI-DA replaces all GSI-DA in the GFS.

RussTreadon-NOAA commented:
**Repository question**
The new job `JGLOBAL_ATMOS_SFCANL` and new script `exglobal_atmos_sfcanl.sh` execute `global_cycle` and `global_cycle.sh`.  Source code for this executable and the script reside in ufs_utils.   

Moving ahead do we want to maintain the surface analysis j-job and script in the [g-w](https://github.com/NOAA-EMC/global-workflow.git) repo or the [ufs_utils](https://github.com/ufs-community/UFS_UTILS.git) repo?

For the time being I will make suggested changes to the sfcanl j-job and ex-script in `feature/ufsda-gdasapp` but please discuss & decide upon the long term home for this j-job and ex-script.   The outcome of this discussion has bearing on other apps executed by the workflow.

Adding @GeorgeGayno-NOAA for his awareness.

RussTreadon-NOAA committed:
Incoporate review feedback for JEDI, postanl, and sfcanl

RussTreadon-NOAA committed:
Remove DOGCYCLE from sfcanl script, correct typo in config.postanl

RussTreadon-NOAA committed:
Merge branch 'develop' at f0f1025 into feature/ufsda-gdasapp

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMOS_SFCANL:
```diff
@@ -62,7 +62,7 @@ export COMPONENT=${COMPONENT:-atmos}
     export ROTDIR=${COMROOT:?}/$NET/$envir
 fi
-export DO_CALC_ANALYSIS=${DO_CALC_ANALYSIS:-"YES"} 
+##export DO_CALC_ANALYSIS=${DO_CALC_ANALYSIS:-"YES"} 
```
Please remove this now-inactive code

RussTreadon-NOAA replied:
Inactive code removed.

WalterKolczynski-NOAA commented on parm/config/config.resources:
```diff
@@ -330,8 +334,12 @@ elif [ $step = "postanl" ]; then
     export npe_node_dwn=$npe_node_max
-    if [[ "$machine" = "WCOSS_DELL_P3" ]]; then export npe_node_postanl=28 ; fi
+    if [[ "$machine" = "WCOSS_DELL_P3" ]]; then
+	export npe_dwn=28
```
This if-block should be indented

RussTreadon-NOAA replied:
It seems these lines are indented in the repo snapshot of [`parm/config/config.resources`](https://github.com/NOAA-EMC/global-workflow/blob/feature/ufsda-gdasapp/parm/config/config.resources#L339)
```
    export npe_node_dwn=$npe_node_max
    if [[ "$machine" = "WCOSS_DELL_P3" ]]; then
	export npe_dwn=28
	export npe_node_postanl=28
    fi
    if [[ "$machine" == "WCOSS_C" ]]; then export memory_postanl="3072M"; fi
```

WalterKolczynski-NOAA replied:
What you pasted isn't indented either. Here's what I see:

![image](https://user-images.githubusercontent.com/28147896/176517675-c273a455-0a25-4bb9-a76c-e4c23aec546b.png)

The two exports are at the same level as the if ... DELL

RussTreadon-NOAA replied:
Interesting.  It's indented on my screen.   The scripting in question is in a `postanl` section.  Reviews have requested removal of `postanl`.  I'm in the process of removing `postatnl` and testing that removal does not break anything.    Assuming successful tests the script fragment in question will be removed and not in the repo.

WalterKolczynski-NOAA replied:
If we can get by without the job too, great, but I was mostly concerned about the duplicate script.

As for this, maybe a space/tab conflict causing different indentation to be shown?

RussTreadon-NOAA replied:
Indentation issue resolved.

WalterKolczynski-NOAA reviewed (changes requested):
Just a couple new minor ones have popped up, plus the outstanding questions about postanal.sh.

RussTreadon-NOAA committed:
Refactor JEDI-VAR and JEDI-ENS jobs, remove postanl (#871)
 - refactor JEDI-VAR and JEDI-ENS rocoto jobs to be consistent with other jobs
 - remove postanl job, config, and xml generation
 - split JEDI-VAR and JEDI-ENS resources into job specific sections
 - link function specific JEDI-VAR and JEDI-ENS j-jobs

RussTreadon-NOAA committed:
add JEDI-VAR and JEDI-ENS j-jobs to .gitignore

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Merge branch 'develop' at 13385d9c into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Merge branch 'develop' at e8361cc into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Merge branch 'develop' at 5a58fa8 into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/ufsda-gdasapp

RussTreadon-NOAA commented:
xml generation has changed in `develop`.   As such, JEDI-VAR and JEDI-ENS xml generation must be added to the new python scripts in `develop/workflow`.  This work is underway in `feature/ufsda-gdasapp`.  This PR will be updated once this development is complete.   Then it's time to re-review this PR and hopefully merge it into `develop`.

RussTreadon-NOAA committed:
Update xml generation to include JEDI_VAR and JEDI_ENS

RussTreadon-NOAA commented:
The commit to `feature/ufsda-gdasapp` at fbbc582c adds JEDI_VAR and JED_ENS scripting to `workflow/applications.py` and `workflow/rocoto/workflow_tasks.py`.   When `workflow/setup_xml.py` is executed from `feature/ufsda-gdasapp` without JEDI_VAR and JEDI_ENS, it reproduces the xml generated by the `develop` branch.

I will run two C96/C48L127 parallels on Orion:
- use GSI based DA to ensure `feature/ufsda-gdasapp` reproduces `develop`
- use JEDI based DA to ensure `feature/ufsda-gdasapp` reproduces previous JEDI DA results

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/ufsda-gdasapp

RussTreadon-NOAA committed:
Update JEDI_VAR dependencies for sfcanl and analcalc

WalterKolczynski-NOAA commented on workflow/applications.py:
```diff
@@ -170,15 +172,27 @@ def _cycled_configs(self):
-                   'fcst', 'post', 'vrfy', 'arch']
+        configs = ['prep']
+##                   'sfcanl', 'analcalc',
+##                   'fcst', 'post', 'vrfy', 'arch']
```
Please remove the inactive lines

RussTreadon-NOAA replied:
Sorry about that.   The inactive lines have been removed.  Done at [594f68d](https://github.com/NOAA-EMC/global-workflow/commit/594f68db15d744f195b33a11170cce4fb4173ef6)

WalterKolczynski-NOAA commented on workflow/applications.py:
```diff
@@ -331,12 +352,19 @@ def _get_cycled_task_names(self):
 
         # Collect all "gdas" cycle tasks
-        gdas_tasks = gdas_gfs_common_tasks_before_fcst + ['analdiag']
+        gdas_tasks = []
+        if self.do_jedivar:
+            gdas_tasks += gdas_gfs_common_tasks_before_fcst
+        else:
+            gdas_tasks += gdas_gfs_common_tasks_before_fcst + ['analdiag']
```
Could this more simply be:
```
gdas_tasks = gdas_gfs_common_tasks_before_fcst
if not self.do_jedivar:
    gdas_tasks += ['analdiag']
```

RussTreadon-NOAA replied:
@WalterKolczynski-NOAA , when I make the suggested change, `setup_xml.py` fails.  Here's an example using config files for GSI-based DA.

```
Orion-login-2[86] rtreadon$ ./setup_xml.py /work/noaa/da/Russ.Treadon/para_gfs/prdev3
sourcing config.prep
sourcing config.anal
sourcing config.analdiag
sourcing config.sfcanl
sourcing config.analcalc
sourcing config.fcst
sourcing config.post
sourcing config.vrfy
sourcing config.arch
sourcing config.gldas
sourcing config.eobs
sourcing config.eomg
sourcing config.ediag
sourcing config.eupd
sourcing config.ecen
sourcing config.esfc
sourcing config.efcs
sourcing config.echgres
sourcing config.epos
sourcing config.earc
sourcing config.metp
Traceback (most recent call last):
  File "./setup_xml.py", line 57, in <module>
    xml = RocotoXML(app_config)
  File "/work/noaa/da/rtreadon/git/global_workflow/update/workflow/rocoto/workflow_xml.py", line 24, in __init__
    task_list = get_wf_tasks(app_config)
  File "/work/noaa/da/rtreadon/git/global_workflow/update/workflow/rocoto/workflow_tasks.py", line 1227, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
  File "/work/noaa/da/rtreadon/git/global_workflow/update/workflow/rocoto/workflow_tasks.py", line 129, in get_task
    return getattr(self, task_name, *args, **kwargs)()
  File "/work/noaa/da/rtreadon/git/global_workflow/update/workflow/rocoto/workflow_tasks.py", line 1107, in echgres
    self._is_this_a_gdas_task(self.cdump, 'echgres')
  File "/work/noaa/da/rtreadon/git/global_workflow/update/workflow/rocoto/workflow_tasks.py", line 64, in _is_this_a_gdas_task
    raise TypeError(f'{task_name} must be part of the "gdas" cycle and not {cdump}')
TypeError: echgres must be part of the "gdas" cycle and not gfs
Orion-login-2[87] rtreadon$
```
The line `gdas_tasks = gdas_gfs_common_tasks_before_fcst` results in `gdas_gfs_common_tasks_before_fcst` being updated when `gdas_tasks` is updated.   I confirmed this by adding `print` statements for `gdas_gfs_common_tasks_before_fcst` after each line in which `gdas_tasks` was updated.

As a result, by the time we get to 
```
        # Collect "gfs" cycle tasks
        gfs_tasks = gdas_gfs_common_tasks_before_fcst
```
`gdas_gfs_common_tasks_before_fcst` contains common gdas_gfs tasks plus gdas only tasks.

I am not famiiliar with python.  Does the assignment `gdas_tasks = gdas_gfs_common_tasks_before_fcst`
mean that these two lists reference the same memory location?  Therefore, updates to `gdas_tasks` also update `gdas_gfs_common_tasks_before_fcst`.

CoryMartin-NOAA replied:
"two lists reference the same memory location? " @RussTreadon-NOAA yes almost always it is a 'shallow copy'.

RussTreadon-NOAA replied:
Thanks, @CoryMartin-NOAA .   Given this, can we simplify

[workflow/applications.py](https://github.com/NOAA-EMC/global-workflow/pull/871/files/abb15aba906e6774265441a9ad854517347bb0a2#diff-d20f9a08d10f570a2399255f8699544a2232eb355efa36a64e094cb7cd790765)
```
        gdas_tasks = []
        if self.do_jedivar:
            gdas_tasks += gdas_gfs_common_tasks_before_fcst
        else:
            gdas_tasks += gdas_gfs_common_tasks_before_fcst + ['analdiag']
```
I thought
```
        # Collect all "gdas" cycle tasks
        gdas_tasks += gdas_gfs_common_tasks_before_fcst
        if not self.do_jedivar:
            gdas_tasks += gdas_gfs_common_tasks_before_fcst + ['analdiag']
```
might work, but python flags this with `UnboundLocalError: local variable 'gdas_tasks' referenced before assignment` ...which kinda makes sense.

CoryMartin-NOAA replied:
```
import copy
gdas_tasks = copy.deepcopy(gdas_gfs_common_tasks_before_fcst)
```
should do the trick

aerorahul replied:
Does this not work?
```
gdas_tasks = gdas_gfs_common_tasks_before_fcst.copy()
```

CoryMartin-NOAA replied:
Oh it probably does yeah

RussTreadon-NOAA replied:
Thank you, @aerorahul and @CoryMartin-NOAA , for your suggestions.  I tried both suggestions.   The resulting xml file winds up with two `task_name` entries for jobs `gdasanal`, `gdassfcanl`, `gdasanalcalc`, and `gdasprep`

RussTreadon-NOAA replied:
@aerorahul and @CoryMartin-NOAA .  I found and corrected the mistake which resulted in double entries for the cited jobs.   The following works

```
        # Collect all "gdas" cycle tasks
        gdas_tasks = gdas_gfs_common_tasks_before_fcst.copy()
        if not self.do_jedivar:
            gdas_tasks += ['analdiag']
```

RussTreadon-NOAA replied:
Update as suggested at [866467e](https://github.com/NOAA-EMC/global-workflow/commit/866467e49b97cb7036f1366193adeac1985fe39c)

WalterKolczynski-NOAA replied:
Yay teamwork!

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -376,6 +384,65 @@ def analdiag(self):
+        gfs_enkf = True if self.app_config.do_hybvar and 'gfs' in self.app_config.eupd_cdumps else False
+
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{"gdas"}post', 'offset': '-06:00:00'}
```
This should be just `'name': 'gdaspost'`, as there's no need to use an f-string if you are just going to put a literal in the { }.

RussTreadon-NOAA replied:
Thanks.  Recommended change made at [594f68d](https://github.com/NOAA-EMC/global-workflow/commit/594f68db15d744f195b33a11170cce4fb4173ef6)

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -376,6 +384,65 @@ def analdiag(self):
+        dump_suffix = self._base["DUMP_SUFFIX"]
+        gfs_cyc = self._base["gfs_cyc"]
+        dmpdir = self._base["DMPDIR"]
+        gfs_enkf = True if self.app_config.do_hybvar and 'gfs' in self.app_config.eupd_cdumps else False
```
Change this variable to `do_gfs_enkf` so it is clear if is a bool.

RussTreadon-NOAA replied:
Recommended change made at [594f68d](https://github.com/NOAA-EMC/global-workflow/commit/594f68db15d744f195b33a11170cce4fb4173ef6)

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -376,6 +384,65 @@ def analdiag(self):
+        dep_dict = {'type': 'task', 'name': f'{self.cdump}atmanalprep'}
+        deps.append(rocoto.add_dependency(dep_dict))
+        if self.app_config.do_hybvar:
+            dep_dict = {'type': 'metatask', 'name': f'{"gdas"}epmn', 'offset': '-06:00:00'}
```
Same as above.

RussTreadon-NOAA replied:
Recommended change made at [39fe881](https://github.com/NOAA-EMC/global-workflow/commit/39fe8813f544713b3205e0a669a04b2060fd8b96)

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -880,6 +947,61 @@ def eupd(self):
+        gfs_enkf = True if self.app_config.do_hybvar and 'gfs' in self.app_config.eupd_cdumps else False
+
+        deps = []
+        dep_dict = {'type': 'metatask', 'name': f'{"gdas"}post', 'offset': '-06:00:00'}
```
Same as above

RussTreadon-NOAA replied:
Recommended change made at [594f68d](https://github.com/NOAA-EMC/global-workflow/commit/594f68db15d744f195b33a11170cce4fb4173ef6)

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -880,6 +947,61 @@ def eupd(self):
+        deps = []
+        dep_dict = {'type': 'task', 'name': f'{self.cdump}atmensanalprep'}
+        deps.append(rocoto.add_dependency(dep_dict))
+        dep_dict = {'type': 'metatask', 'name': f'{"gdas"}epmn', 'offset': '-06:00:00'}
```
Same as above.

RussTreadon-NOAA replied:
Recommended change made at [594f68d](https://github.com/NOAA-EMC/global-workflow/commit/594f68db15d744f195b33a11170cce4fb4173ef6)

WalterKolczynski-NOAA reviewed (changes requested):
Mostly minor

RussTreadon-NOAA committed:
Update JEDI_ENS dependencies for ecen and esfc

RussTreadon-NOAA committed:
Update workflow python scripts as recommended by reviewer

RussTreadon-NOAA committed:
Update atmanalrun dependency as suggested by reviewer

RussTreadon-NOAA committed:
Update initialization of gdas_tasks as suggested by reviewers

RussTreadon-NOAA commented:
The GSI and JEDI based DA parallels have completed two cycles on Orion.  Results are consistent with previous runs.

WalterKolczynski-NOAA commented:
@RussTreadon-NOAA if you can add a bit more to the PR description about what changed and usage, I can merge this tonight.

WalterKolczynski-NOAA commented:
> @RussTreadon-NOAA if you can add a bit more to the PR description about what changed and usage, I can merge this tonight.

Usage being how to invoke it, not necessarily how to actually use it in a research sense.

RussTreadon-NOAA committed:
Add flag for JEDI_ENS to config.base.nco.static

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA , the description has been updated.  Is this sufficient?

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA , the description has been updated. Is this sufficient?

Yes, excellent. Thanks Russ. This will be merged shortly.

WalterKolczynski-NOAA commented:
@RussTreadon-NOAA oh, one last thing. Does this actually close #521?

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA .   Yes, please close #521 with this PR.    Issue #521 focuses on atmospheric JEDI DA.   This PR adds the infrastructure for prototype atmospheric JEDI DA to the global workflow.   Once the machinery is in place we can move onto the science.
